Sources/ComposableArchitecture/Internal/CurrentValueRelay.swift

```diff
@@ -14,42 +15,128 @@ final class CurrentValueRelay<Output>: Publisher {
-    self.subscriptions.append(subscription)
+  deinit {
+    self.lock.deinitialize(count: 1)
+    self.lock.deallocate()
+  }
+
+  func receive(subscriber: some Subscriber<Output, Never>) {
+    let subscription = Subscription(upstream: self, downstream: subscriber)
+    self.lock.sync {
+      self.subscriptions.append(subscription)
+    }
     subscriber.receive(subscription: subscription)
-    subscription.forwardValueToBuffer(self.currentValue)
   }
 
   func send(_ value: Output) {
     self.currentValue = value
-    for subscription in subscriptions {
-      subscription.forwardValueToBuffer(value)
+    for subscription in self.subscriptions {
+      subscription.receive(value)
+    }
+  }
+
+  private func remove(_ subscription: Subscription) {
+    self.lock.sync {
+      guard let index = self.subscriptions.firstIndex(of: subscription)
+      else { return }
+      self.subscriptions.remove(at: index)
     }
   }
 }
 
 extension CurrentValueRelay {
-  final class Subscription<Downstream: Subscriber>: Combine.Subscription
-  where Downstream.Input == Output, Downstream.Failure == Failure {
-    private var demandBuffer: DemandBuffer<Downstream>?
+  fileprivate final class Subscription: Combine.Subscription, Equatable {
+    private var demand = Subscribers.Demand.none
+    private var downstream: (any Subscriber<Output, Never>)?
+    private let lock: os_unfair_lock_t
+    private var receivedLastValue = false
+    private var upstream: CurrentValueRelay?
 
-    init(downstream: Downstream) {
-      self.demandBuffer = DemandBuffer(subscriber: downstream)
+    init(upstream: CurrentValueRelay, downstream: any Subscriber<Output, Never>) {
+      self.upstream = upstream
+      self.downstream = downstream
+      self.lock = os_unfair_lock_t.allocate(capacity: 1)
+      self.lock.initialize(to: os_unfair_lock())
     }
 
-    func forwardValueToBuffer(_ value: Output) {
-      _ = demandBuffer?.buffer(value: value)
+    deinit {
+      self.lock.deinitialize(count: 1)
+      self.lock.deallocate()
+    }
+
+    func cancel() {
+      self.lock.sync {
+        self.downstream = nil
+        self.upstream?.remove(self)
+        self.upstream = nil
+      }
+    }
+
+    func receive(_ value: Output) {
+      guard let downstream else { return }
+
+      switch self.demand {
+      case .unlimited:
+        // NB: Adding to unlimited demand has no effect and can be ignored.
+        _ = downstream.receive(value)
+
+      case .none:
+        self.lock.sync {
+          self.receivedLastValue = false
+        }
+
+      default:
+        self.lock.sync {
+          self.receivedLastValue = true
+          self.demand -= 1
+        }
+        let moreDemand = downstream.receive(value)
+        self.lock.sync {
+          self.demand += moreDemand
+        }
+      }
     }
 
     func request(_ demand: Subscribers.Demand) {
-      _ = demandBuffer?.demand(demand)
+      precondition(demand > 0, "Demand must be greater than zero")
```

Comment:
This is a breaking change, however, it reflects the behaviour of `CurrentValueSubject`. Also, I think requesting zero demand would be considered a programmer error.